[Blog] Add post on HTTP cache and security (#7280)
GoogleChrome/web.dev: src/site/content/en/blog/http-cache-security/index.md

gain unauthorized access to cross-origin data. As a consequence, modern web
browsers have restricted usage of some of their features&mdash;such as
[`SharedArrayBuffer`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer)
or [high resolution timer](https://developer.chrome.com/blog/cross-origin-isolated-hr-timers/)&mdash;to
pages with [cross-origin
isolation](/cross-origin-isolation-guide/). 

Modern web browsers enforce [Cross-Origin Embedder Policy
(COEP)](https://developer.chrome.com/blog/coep-credentialless-origin-trial/). This ensures cross-origin
resources are either:

*  Public resources, requested without cookies
*  Resources explicitly allowed to be shared cross-origin, via CORS or the
   [CORP](https://developer.mozilla.org/docs/Web/HTTP/Cross-Origin_Resource_Policy_(CORP))
   header

This setup doesn't prevent an attacker from exploiting Spectre. However, it
ensures cross-origin resources are not valuable to the attacker (= public
resource) or allowed to be shared with the attacker (=

> `[CORP](https://developer.mozilla.org/docs/Web/HTTP/Cross-Origin_Resource_Policy_(CORP)):cross-origin`).

Reviewer: Apparently it's not possible to put a partial link in a code block.

```suggestion
[`CORP: cross-origin`](https://developer.mozilla.org/docs/Web/HTTP/Cross-Origin_Resource_Policy_(CORP))).
```

### How does HTTP caching affect Spectre?

If the `Cache-Control` header isn't properly set, an attacker could execute an
attack. For example:

1.  The credentialed resource is cached.
1.  The attacker loads a cross-origin isolated page.
1.  The attacker requests the resource again.

> 1.  `COEP:credentialless` is set by the browser, so the resource is fetched

Reviewer: Perhaps we can add a link to https://developer.chrome.com/blog/coep-credentialless-origin-trial/ ?

## Common misconceptions about HTTP caches

### 1. Resources are cached by the browser only

There are often multiple layers of cache. Some caches are dedicated to a
single user, some to multiple users. Some are controlled by the server, some
by the user, and some by intermediaries.

*  **Browser caches**. These caches are owned by and dedicated to a single
   user, implemented in their web browser. They improve performance by
   avoiding fetching the same response multiple times.
*  **Local proxy**. This might have been installed by the user, but can also
   be managed by intermediaries: their company, their organization, or their
   internet provider. Local proxies often cache a single response for multiple
   users, which constitutes a "public" cache. Local proxies have multiple
   roles.
*  **Origin server cache / CDN**. This is controlled by the server. The Origin
   server cache's goal is to reduce the load on the origin server by caching
   the same response for multiple users. A CDN's goals are similar, but spread
   across the globe and assigned to the closest set of users to reduce latency.

<figure data-size="full">
  {% Img
     class="screenshot",

>      src="image/VbsHyyQopiec0718rMq2kTE1hke2/2VVXjpJLgTwXiPEFTcAI.jpg",

Reviewer: Do we have a higher resolution image of this? I'd replace.

Reviewer 2: Here is an svg:
https://drive.google.com/file/d/195uEmhdHpSgZWrZ1gFLfLK_HSQtsXWSp/view?usp=sharing
![HTTP Cache   Security Blog Post (1)](https://user-images.githubusercontent.com/4759106/151937899-3f4d7d37-ce8c-4b3a-8357-fe667685211e.svg)

Author: Great thank you -- the SVG was added and uploaded. The preview should update soon.

     alt="There are multiple layers of cache between the browser and server.",
     width="380", height="240"
  %}

>   <figcaption>There are various layers of cache between the browser and server. This begins with a public server cache, followed by a CDN, and the browser cache. There may also be a local proxy setup between the CDN and browser cache.</figcaption>

Reviewer 2: Technically, the server's cache and its CDN are also optional. It might worth saying "there may be be" instead of "there are"
```
There are various layers of cache between the browser and server. There may be the server's own cache, its CDN, and finally the browser's cache. There may also be a local proxy setup between the CDN and browser cache.
```

Note: I also removed "public" from the "a public server cache". I was not sure about it.

---
layout: post
title: Improve security and privacy by updating HTTP Cache
authors:
  - arthursonzogni
description: Forgetting or misusing the Cache-Control header might negatively impact the security of your website and your users' privacy. Get recommendations for high-value websites.

> date: 2022-02-07

Reviewer: Anytime. Feb 7th sounds fine.

thumbnail: image/VbsHyyQopiec0718rMq2kTE1hke2/3MnHBfutFgu2ai1F2mUH.jpg
alt: Two ducks move through the water, creating a rippling wave.
tags:
  - blog
  - security
---

By default, resources are always allowed to be cached by any type of cache.
Not using or misusing the `Cache-Control` header might negatively impact the
security of your website and your users' privacy.

For personalized responses you want to keep private, we recommend you either:

*  Set an appropriate [secondary cache
   key](https://httpwg.org/specs/rfc7234.html#caching.negotiated.responses).
   If the response depends on cookies, then set `Vary: Cookie`.
*  Prevent intermediaries from caching the resource. Set

>    `Cache-Control: private`.

Reviewer 2: @JackJey suggested putting the emphasis on the second point and making it clearer on the first point that `Vary:Cookie` can be used only when the Cookie is the credentials.

Maybe something like:
```
*  Prevent intermediaries from caching the resource. Set
   `Cache-Control: private`.
*  Set an appropriate [secondary cache
   key](https://httpwg.org/specs/rfc7234.html#caching.negotiated.responses).
   If the response varies upon cookie, because it stores credentials, then set `Vary: Cookie`.
```

Here and in L150.

WDYT @JackJey ?

Author: I've updated as follows: 

> *  Prevent intermediaries from caching the resource. Set `Cache-Control: private`.
> *  Set an appropriate [secondary cache key]https://httpwg.org/specs/rfc7234.html#caching.negotiated.responses). If the response varies due to cookies&mdash;which can happen when the cookie stores credentials&mdash;then set `Vary: Cookie`.

---
layout: post
title: Improve security and privacy by updating HTTP Cache
authors:
  - arthursonzogni
description: Forgetting or misusing the Cache-Control header might negatively impact the security of your website and your users' privacy. Get recommendations for high-value websites.
date: 2022-02-07
thumbnail: image/VbsHyyQopiec0718rMq2kTE1hke2/3MnHBfutFgu2ai1F2mUH.jpg
alt: Two ducks move through the water, creating a rippling wave.
tags:
  - blog
  - security
---

By default, resources are always allowed to be cached by any type of cache.
Not using or misusing the `Cache-Control` header might negatively impact the
security of your website and your users' privacy.

For personalized responses you want to keep private, we recommend you either:

*  Prevent intermediaries from caching the resource. Set
   `Cache-Control: private`.
*  Set an appropriate [secondary cache

>    key](https://httpwg.org/specs/rfc7234.html#caching.negotiated.responses).

Author: ```suggestion
   key](https://datatracker.ietf.org/doc/html/rfc7234#section-4.1).
```

## Cache-related security and privacy risks

### Leaky resources from Spectre vulnerabilities

The [Spectre
vulnerability](https://en.wikipedia.org/wiki/Spectre_(security_vulnerability))
allows a page to read an OS process's memory. This means an attacker can
gain unauthorized access to cross-origin data. As a consequence, modern web
browsers have restricted usage of some of their features&mdash;such as
[`SharedArrayBuffer`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer)
or [high resolution timer](https://developer.chrome.com/blog/cross-origin-isolated-hr-timers/)&mdash;to
pages with [cross-origin
isolation](/cross-origin-isolation-guide/). 

Modern web browsers enforce [Cross-Origin Embedder Policy
(COEP)](https://developer.chrome.com/blog/coep-credentialless-origin-trial/). This ensures cross-origin
resources are either:

*  Public resources, requested without cookies
*  Resources explicitly allowed to be shared cross-origin, via CORS or the
   [CORP](https://developer.mozilla.org/docs/Web/HTTP/Cross-Origin_Resource_Policy_(CORP))
   header

The COEP setup doesn't prevent an attacker from exploiting Spectre. However, it

> ensures cross-origin resources are not valuable to the attacker (when shared

Reviewer 2: In the first case, the action is made by the browser, by not sending cookies. So I would rather say "when loaded as").
In the second case, the action is made by the server, so I would keep "when shared with"

Author: Done!

> ensures cross-origin resources are not valuable to the attacker (when shared

Author: ```suggestion
ensures cross-origin resources are not valuable to the attacker (when loaded as
```

> as public resource) or allowed to be shared with the attacker (when shared with

Author: ```suggestion
by the browser as public resource) or allowed to be shared with the attacker (when shared with
```

     alt="There are often multiple layers of cache between the browser and server.",
     width="800", height="712"
  %}
  <figcaption>
     There may be various layers of cache between the browser and server. For example, you may encounter a server cache, followed by a CDN and the browser cache. There may also be a local proxy setup between the CDN and browser cache.
  </figcaption>
</figure>

### 2. SSL prevents intermediaries from caching HTTPS resources

Many users regularly use locally-configured proxies, whether for access
purposes (such as sharing a metered connection), virus inspection, or for data
loss prevention (DLP) purposes. The intermediary cache is performing [TLS
interception](https://en.wikipedia.org/wiki/Transport_Layer_Security#TLS_interception). 

An intermediary cache is often installed on a company employee's workstations.
Web browsers are configured to trust the local proxy's certificates.

Ultimately, some HTTPS resources may be cached by these local proxies.

## How HTTP cache works

*  Resources are
   [implicitly](https://datatracker.ietf.org/doc/html/rfc7234) allowed
   to be cached by default.
*  The [primary cache

>    key](https://httpwg.org/specs/rfc7234.html#rfc.section.2) consists of the

Author: ```suggestion
   key](https://datatracker.ietf.org/doc/html/rfc7234#section-2) consists of the
```
